iot-e2e-tests/common/src/main/java/com/microsoft/azure/sdk/iot/common/helpers/CorrelationDetailsLoggingAssert.java

```diff
@@ -0,0 +1,260 @@
+                }
+
+                isFirstDevice = false;
+            }
+        }
+
+        if (moduleIds.size() > 0)
+        {
+            correlationString += " Module id: ";
+            boolean isFirstModule = true;
+            for (String moduleId : moduleIds)
+            {
+                if (isFirstModule)
+                {
+                    correlationString += moduleId;
+                }
+                else
+                {
+                    correlationString += ", " + moduleId;
+                }
+
+                isFirstModule = false;
+            }
+        }
+
+        correlationString = correlationString + " Protocol: " + protocol + " Timestamp: " + timeStamp;
+
+        return baseMessage + "\n(Correlation details: <" + correlationString + ">)";
+    }
+
+    public static String BuildExceptionMessage(String baseMessage, String deviceId, String protocol, String hostname, String moduleId)
+    {
+        Collection<String> deviceIds = new ArrayList<>();
+        deviceIds.add(deviceId);
+
+        Collection<String> moduleIds = new ArrayList<>();
+        if (moduleId != null && !moduleId.isEmpty())
+        {
+            moduleIds.add(moduleId);
+        }
+
+        return BuildExceptionMessage(baseMessage, deviceIds, protocol, hostname, moduleIds);
+    }
+
+    public static String BuildExceptionMessage(String baseMessage, InternalClient client)
+    {
+        if (client == null || client.getConfig() == null)
+        {
+            throw new IllegalArgumentException("client and config must not be null");
+        }
+
+        return BuildExceptionMessage(
+                baseMessage,
+                client.getConfig().getDeviceId(),
+                client.getConfig().getProtocol().toString(),
+                client.getConfig().getIotHubHostname(),
+                client.getConfig().getModuleId());
+    }
+
+    public static String BuildExceptionMessage(String baseMessage, Collection<InternalClient> clients)
+    {
+        String hostname = "";
+        String protocol = "";
+        Collection<String> deviceIds = new ArrayList<>();
+        Collection<String> moduleIds = new ArrayList<>();
+
+        for (InternalClient client : clients)
+        {
+            hostname = client.getConfig().getIotHubHostname();
+            protocol = client.getConfig().getProtocol().toString();
+            deviceIds.add(client.getConfig().getDeviceId());
+
+            if (client.getConfig().getModuleId() != null && !client.getConfig().getModuleId().isEmpty())
+            {
+                moduleIds.add(client.getConfig().getModuleId());
+            }
+        }
+
+        return BuildExceptionMessage(
+                baseMessage,
+                deviceIds,
+                protocol,
+                hostname,
+                moduleIds);
+    }
+
+    public void assertTrue(String message, boolean condition)
+    {
+        if (!condition)
+        {
+            fail(message);
+        }
+    }
```

Comment:
why can't we use libraries to acheive this ? Why do we need to write this method?

Reply:
These are instance methods that make use of the instance variables provided in the constructor. This assertTrue calls into fail() which builds a better failure message than the regular Assert.fail() would provide on its own